Introduce `tanzu local` model and deprecate `tanzu standalone-cluster`

joshrosso opened the pull request:
**Warning**:  The issue this PR correlates to (https://github.com/vmware-tanzu/community-edition/issues/2266) is under review. We have not committed to this functionality at this point. For discussion around the proposal **please** do this inside the issue and keep PR comments to the implementation details. Thank you!

This PR implements `tanzu local` as an alternative to `standalone-clusters`.

### Testing it:

Right now, several assumptions are made about your environment. Namely that you have tanzu (0.9.0+) pre-installed. If so, you can run the following to try out `tanzu local`

1. Be sure to run all command **from the root** of this repository

    ```sh
     make build-tce-cli-plugins ENVS=linux-amd64 && cp -v ./artifacts/local/v0.10.0-dev.1/tanzu-local-linux_amd64 ~/.local/share/tanzu-cli/tanzu-plugin-local
    ```
    
    > If not on `linux`, change the above to point to your `$XDG_DATA_HOME`

1. Run `tanzu local`
    
    ```
    $ tanzu local
    Manage local environments of Tanzu
    
    Usage:
      tanzu local [command]
    
    Available Commands:
      completion    generate the autocompletion script for the specified shell
      create        create a local tanzu cluster
      delete        delete a local tanzu cluster
      list          list tanzu clusters
    
    Flags:
      -h, --help              help for local
          --log-file string   Log file path
      -v, --verbose int32     Number for the log level verbosity(0-9)
    
    Use "tanzu local [command] --help" for more information about a command.
    ```


### Cluster creation:

![1](https://user-images.githubusercontent.com/6200057/137748829-f63968e3-1534-42a4-a3a1-707e6f063f1a.gif)

### Cluster init:

![2](https://user-images.githubusercontent.com/6200057/137748851-171caeff-600f-4f0d-9781-6eefffe71b0b.gif)

### Cluster list / deletion:

![3](https://user-images.githubusercontent.com/6200057/137748872-ffb5de7b-5c7a-4a53-a5ea-3a9d607b92cc.gif)

Resolves: https://github.com/vmware-tanzu/community-edition/issues/2266

Commit:
Initial commit for local clusters

This commit provides local cluster functionality to the tanzu-cli. It
also replaces the existing standalone-cluster plugin. As of this commit
it does work, but makes several assumptions about the environment.

jorgemoralespou commented:
>> make build-tce-cli-plugins ENVS=linux-amd64 && cp -v ./artifacts/local/v0.10.0-dev.1/tanzu-local-linux_amd64 ~/.local/share/tanzu-cli/tanzu-plugin-local

>If not on linux, change the above to point to your $XDG_DATA_HOME

nrb commented on cli/cmd/plugin/standalone-cluster/create.go:
```diff
@@ -5,17 +5,42 @@ package main
 
 import (
 	"fmt"
+	"io/ioutil"
```
I think this directory should be renamed, for clarity.

nrb commented on cli/cmd/plugin/standalone-cluster/create.go:
```diff
@@ -54,103 +83,374 @@ func create(cmd *cobra.Command, args []string) error {
+					return err
+				}
+
+				// TODO(jpmcb): in the future, maybe we consider using cliient-go to actually create the objects instead of shilling out to kubectl
```
+1 to client-go.

jpmcb committed:
Implements pulling kapp image bundle and parsing it with ytt

Signed-off-by: John McBride <jmcbride@vmware.com>

nrb reviewed (commented):
Looks like `cli/cmd/plugin/standalone-cluster/standalone-cluster` is a committed binary, not sure if you wanted to do that.

Also, is there a definition of the provider interface we're aiming to create as mentioned in the issue yet? I'm not seeing one by searching for `interface` if so.

stmcginnis committed:
Remove ethtool containers when done

During cluster deployment, the ethtool container is run a few times to
get network information. These containers are run by calling out to
docker run, then they are left behind after the output is received from
them.

This adds the `--rm` argument to the `docker run` command so these
containers go away after we are done with them.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

stmcginnis committed:
Add cluster module to local plugin

This adds an abstraction for working with clusters to the local plugin.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

jpmcb committed:
Implements the TkrImageReader interface

enables operations on an individual bundle image.

Signed-off-by: John McBride <jmcbride@vmware.com>

Commit:
Merge pull request #3 from jpmcb/standalone-overhaul

Implement `TkrImageReader`

Commit:
Implement kapp package

This package is reposible for installing and providing the status of
kapp-controller

Signed-off-by: joshrosso <rossoj@vmware.com>

stmcginnis committed:
Start to refactor kind calls into interface

This moves some of the cluster interface into a kind specific
implementation.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

Commit:
Merge pull request #1 from stmcginnis/sa-overhaul-ethtool

Remove ethtool containers when done

stmcginnis committed:
Add ability to list nodes

This completes removal of direct kind interaction from the command code
and routes everything through the cluster manager abstraction.

Also includes minor cleanup.

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

stmcginnis committed:
Various code cleanup

Signed-off-by: Sean McGinnis <smcginnis@vmware.com>

Commit:
Merge pull request #2 from stmcginnis/sa-overhaul-cluster

Add cluster module to local plugin
